### PR TITLE
Unit Testing

### PR DESCRIPTION
**Pull request by takanokage**
#166 

This is a skeleton implementation of how Google Test can be used to implement unit testing.
A proof of concept is provided through `src/UnitTest/Image_unittest.cpp`. This example is not meant to fully test the `Core\Geometry\Image` class, only to show how it's done.

Brief documentation is provided on the `Compilation Options` page.

Please take a look and let me know if you have improvement suggestions.

### CODE REVIEW

**Comment by takanokage** [comment]
There's something wrong with the master branch and/or the AppVeyor script.
By default the unit tests are disabled.

**Comment by syncle** [comment]
Yes. The cure can be found from #317. I halted merging yesterday but you can make a PR again.

**Comment by takanokage** [comment]
Ok, no problem.

**Comment by takanokage** [comment]
This can be a squash merge once we're ready.

File: src/UnitTest/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+cmake_minimum_required(VERSION 3.0)
+
+# Locate GTest
+find_package(GTest REQUIRED)
+include_directories(${GTEST_INCLUDE_DIRS})
+
+# Specify the unit tests
+add_executable(unitTests
+               main.cpp
+               Image_unittest.cpp)
```
Comment (syncle) [inline src/UnitTest/CMakeLists.txt]: I think `image.cpp` would be fine, as it is under UnitTest folder.

**Comment by syncle** [comment]
Can you open PR for AppVeyor issue too? Thanks!

File: src/UnitTest/CMakeLists.txt (hunk 1 above)
Comment (takanokage) [reply src/UnitTest/CMakeLists.txt]: yeah, we could do that.

**Comment by takanokage** [comment]
Actually I haven't worked on the AppVeyor at all so far. I don't know what that issue is about.
Expect a ramp up period (and subsequently no progress on other things) if you really want me to help with that.

**Comment by takanokage** [comment]
I just added skeleton unit test code for most of the features in the three libraries: Core, IO & Visualization.
By default each case displays a "Not implemented." message in red. I could use the FAIL() macro to actually cause a fail by default in each test case however the output of the unit test becomes quite inelegant (output becomes very verbose which seems overkill for the purpose of just signaling that the test case hasn't been implemented yet.)

The output will look like this:

```
[----------] 1 test from PinholeCameraTrajectoryIO
[ RUN      ] PinholeCameraTrajectoryIO.Default
[          ] Not implemented.
[       OK ] PinholeCameraTrajectoryIO.Default (0 ms)
[----------] 1 test from PinholeCameraTrajectoryIO (0 ms total)
```

There might be cases where it doesn't make sense to ever add tests. If that is the case we can just delete the offending unit test file.

Oh, and by the way I maintained the same directory structure as was found under Core, IO and Visualization.

**Comment by takanokage** [comment]
I've also been investigating how to get the unit tests running on Windows.
So far I figured out the google test installation part. However `find_package(GTest REQUIRED)` fails on Windows and I'm not sure there's anything to do about it. If we really want unit tests on Windows we might have to provide a manual path for including and linking against the gtest.

**Comment by syncle** [comment]
@takanokage: checkout this https://github.com/iat-cener/tonatiuh/wiki/Installing-Google-Test-For-Windows

**Comment by takanokage** [comment]
Yeah, I saw that link. As it happens that didn't help. I've gone past the gtest installation, it's the cmake automation for finding gtest libs and headers that fails.

**Comment by takanokage** [comment]
Question: do we want to add Windows support as a separate PR or as part of this PR?

**Comment by takanokage** [comment]
I propose we make it a separate PR because the Windows side has some issues that will take time to address. We should also be fine with the current solution which is supported on both macOS and Linux.

**Comment by syncle** [comment]
Can you test with CI with unit test flag on? I can merge this if it passes build. @qianyizh can you briefly review this PR too?

**Comment by takanokage** [comment]
Great, we agree on all points (including `Win32`).
I'm already on it.

**Comment by takanokage** [comment]
gtest required in order to build the unit tests.
I'm working on a script to install that as part of the CI tests.

**Comment by takanokage** [comment]
Let's disconnect this pull request. The new requirement to test unit test building in Travis CI has complicated things.

**Pull request closed without merging**